subprojects/docs/src/docs/userguide/developing-plugins/kotlin_gradle_plugins.adoc

```diff
@@ -0,0 +1,334 @@
+
+TIP: Because it is provided by Gradle at runtime, `gradleKotlinDsl()` must be a `compileOnly` dependency.
+
+[[sec:kotlin_stdlib]]
+=== kotlin-stdlib
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:$version")
+}
+----
+
+The Kotlin compatibility matrix can be found in the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[Gradle compatibility page]. If you want your plugin to be compatible with a given version of Gradle, you must use a compatible version of kotlin-stdlib.
+For example, if you need your plugin to be compatible with Gradle 7.0, you cannot use any Kotlin 1.5 only API (like https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/lowercase.htmllowercase) and need to compile against `kotlin-stdlib:1.4.31`.
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:1.4.31")
+}
+----
+
+For similar reasons, any transitive dependency of your plugin must use a `kotlin-stdlib` version compatible with the Gradle one.
+While you can ensure that at compile time for your code, it's often more complicated for dependencies, so test your plugins and/or double-check your transitive dependencies.
+
+Note: The `kotlin-stdlib` version is independent of your Kotlin Gradle Plugin version. You can use a newer Kotlin Gradle Plugin version with an older `kotlin-stdlib` using the https://kotlinlang.org/docs/compiler-reference.html#api-version-version[`apiVersion`] compiler argument.
+
+TIP: Because it is provided by Gradle at runtime, `kotlin-stdlib` must be a `compileOnly` dependency.
+
+[[sec:kotlin_reflect]]
+=== kotlin-reflect
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-reflect:$version")
+}
+----
+
+Not all plugins need `kotlin-reflect` but if yours does, the situation is the same as for `kotlin-stdlib`. `kotlin-reflect` needs to be compatible with the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[target Gradle version]
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+[[sec:others]]
+=== others (`implementation` or `api`)
+
+You can add other non-compileOnly dependencies like in other JVM projects:
+
+[source,kotlin]
+----
+dependencies {
+    // Kotlin dependencies
+    // ⚠️ make sure they do not depend on an incompatible version of kotlin-stdlib
+    implementation("com.squareup:kotlinpoet:1.12.0")
+
+    // Java dependencies
+    implementation("asm:asm:3.3.1")
+}
+----
+
+Gradle does not provide these dependencies at runtime and should use the `implementation` configuration (or `api` if part of your plugin API). Because Gradle may load different plugins with conflicting dependencies in the same https://dev.to/autonomousapps/build-compile-run-a-crash-course-in-classpaths-f4g[classloader], your plugin users may experience crashes if symbols are incompatible or missing. To mitigate those, you can relocate your dependencies.
```

Comment:
```suggestion
Gradle does not provide these dependencies at runtime and should use the `implementation` configuration (or `api` if part of your plugin API). 
Because Gradle may load different plugins with conflicting dependencies in the same https://dev.to/autonomousapps/build-compile-run-a-crash-course-in-classpaths-f4g[classloader], your plugin users may experience crashes if symbols are incompatible or missing. 
To mitigate this, you can relocate your dependencies.
```